core/app/models/spree/promotion_handler/free_shipping.rb

```diff
@@ -2,41 +2,57 @@ module Spree
   module PromotionHandler
     # Used for activating promotions with shipping rules
     class FreeShipping
-      attr_reader :order, :order_promo_ids
+      attr_reader :order
       attr_accessor :error, :success
 
       def initialize(order)
         @order = order
-        @order_promo_ids = order.promotions.pluck(:id)
       end
 
       def activate
-        promotions.each do |promotion|
-          next if promotion.codes.any? && !order_promo_ids.include?(promotion.id)
+        connected_promotions.each do |order_promotion|
+          order_promotion.promotion.activate(
+            order: order,
+            promotion_code: order_promotion.promotion_code,
+          )
+        end
 
-          if promotion.eligible?(order)
-            promotion.activate(order: order)
-          end
+        not_connected_automatic_promotions.each do |promotion|
+          promotion.activate(order: order)
         end
       end
 
       private
 
-      def promotions
-        promo_table = Promotion.arel_table
-        code_table  = PromotionCode.arel_table
-
-        promotion_code_join = promo_table.join(code_table, Arel::Nodes::OuterJoin).on(
-          promo_table[:id].eq(code_table[:promotion_id])
-        ).join_sources
-
-        Spree::Promotion.active.
-          joins(promotion_code_join).
-          where({
-            id: Spree::Promotion::Actions::FreeShipping.pluck(:promotion_id), # This would probably be more efficient by joining instead
-            spree_promotion_codes: { id: nil },
-            path: nil
-          })
+      def not_connected_automatic_promotions
+        automatic_promotions - connected_promotions.map(&:promotion)
+      end
+
+      def automatic_promotions
+        @automatic_promotions ||= active_free_shipping_promotions.
+          where(apply_automatically: true).
+          to_a.
+          uniq
+      end
+
+      def connected_promotions
+        @connected_promotions ||= order.order_promotions.
+          joins(:promotion).
+          includes(:promotion).
+          merge(active_free_shipping_promotions).
+          to_a.
+          uniq
+      end
+
+      def active_free_shipping_promotions
```

Comment:
what is the SQL generated by this? Think this could be creating some problems for us.